File: compatibility-tests/sdk-test-node/tests/bedrock.test.ts
import { afterAll, beforeAll, describe, expect, it } from 'vitest';
import {
  BedrockClient,
  CreateModelInvocationJobCommand,
  GetModelInvocationJobCommand,
  ListModelInvocationJobsCommand,
  StopModelInvocationJobCommand,
} from '@aws-sdk/client-bedrock';
import {
  S3Client,
  CreateBucketCommand,
  DeleteBucketCommand,
  DeleteObjectCommand,
  GetObjectCommand,
  ListObjectsV2Command,
  PutObjectCommand,
} from '@aws-sdk/client-s3';
import { makeClient, uniqueName } from './setup';

describe('Bedrock Batch Inference', () => {
  let bedrock: BedrockClient;
  let s3: S3Client;
  let bucketName: string;

  beforeAll(async () => {
    bedrock = makeClient(BedrockClient);
    s3 = makeClient(S3Client);
    bucketName = uniqueName('bedrock-bucket');
    await s3.send(new CreateBucketCommand({ Bucket: bucketName }));
  });

  afterAll(async () => {
    if (!bucketName) return;
    try {
      const list = await s3.send(new ListObjectsV2Command({ Bucket: bucketName }));
      for (const obj of list.Contents ?? []) {
        if (obj.Key) {
          await s3.send(new DeleteObjectCommand({ Bucket: bucketName, Key: obj.Key })).catch(() => undefined);
        }
      }
      await s3.send(new DeleteBucketCommand({ Bucket: bucketName })).catch(() => undefined);
    } catch {
      // ignore cleanup errors
    }
  });

  it('creates and retrieves a model invocation job', async () => {
    const jobName = uniqueName('batch-job');
    const inputUri = `s3://${bucketName}/in.jsonl`;
    const outputUri = `s3://${bucketName}/out`;

    const createRes = await bedrock.send(
      new CreateModelInvocationJobCommand({
        jobName,
        modelId: 'amazon.titan-text-express-v1',
        roleArn: 'arn:aws:iam::000000000000:role/BedrockBatchRole',
        inputDataConfig: {
          s3InputDataConfig: {
            s3Uri: inputUri,
          },
        },
        outputDataConfig: {
          s3OutputDataConfig: {
            s3Uri: outputUri,
          },
        },
        tags: [{ key: 'Environment', value: 'testing' }],
      })
    );

    expect(createRes.jobArn).toBeDefined();
    expect(createRes.jobArn).toContain(':model-invocation-job/');

    const getRes = await bedrock.send(
      new GetModelInvocationJobCommand({
        jobIdentifier: createRes.jobArn,
      })
    );

    expect(getRes.jobArn).toBe(createRes.jobArn);
    expect(getRes.jobName).toBe(jobName);
    expect(getRes.modelId).toBe('amazon.titan-text-express-v1');
    expect(getRes.status).toBeDefined();
    expect(getRes.inputDataConfig?.s3InputDataConfig?.s3Uri).toBe(inputUri);
    expect(getRes.outputDataConfig?.s3OutputDataConfig?.s3Uri).toBe(outputUri);
  });

  it('lists model invocation jobs with filtering', async () => {
    const jobName1 = uniqueName('list-job-1');
    const jobName2 = uniqueName('list-job-2');

    await bedrock.send(
      new CreateModelInvocationJobCommand({
        jobName: jobName1,
        modelId: 'amazon.titan-text-express-v1',
        roleArn: 'arn:aws:iam::000000000000:role/BedrockBatchRole',
        inputDataConfig: { s3InputDataConfig: { s3Uri: `s3://${bucketName}/1.jsonl` } },
        outputDataConfig: { s3OutputDataConfig: { s3Uri: `s3://${bucketName}/out1` } },
      })
    );

    await bedrock.send(
      new CreateModelInvocationJobCommand({
        jobName: jobName2,
        modelId: 'anthropic.claude-3-sonnet-20240229-v1:0',
        roleArn: 'arn:aws:iam::000000000000:role/BedrockBatchRole',
        inputDataConfig: { s3InputDataConfig: { s3Uri: `s3://${bucketName}/2.jsonl` } },
        outputDataConfig: { s3OutputDataConfig: { s3Uri: `s3://${bucketName}/out2` } },
      })
    );

    const listRes = await bedrock.send(new ListModelInvocationJobsCommand({}));
    expect(listRes.invocationJobSummaries).toBeDefined();
    const names = listRes.invocationJobSummaries!.map((j) => j.jobName);
    expect(names).toContain(jobName1);
    expect(names).toContain(jobName2);

    const filteredRes = await bedrock.send(
      new ListModelInvocationJobsCommand({
        nameContains: jobName1,
      })
    );
    const filteredNames = filteredRes.invocationJobSummaries!.map((j) => j.jobName);
    expect(filteredNames).toContain(jobName1);
    expect(filteredNames).not.toContain(jobName2);
  });

  it('stops an in-flight model invocation job', async () => {
    const jobName = uniqueName('stop-job');
    const createRes = await bedrock.send(
      new CreateModelInvocationJobCommand({
        jobName,
        modelId: 'amazon.titan-text-express-v1',
        roleArn: 'arn:aws:iam::000000000000:role/BedrockBatchRole',
        inputDataConfig: { s3InputDataConfig: { s3Uri: `s3://${bucketName}/stop.jsonl` } },
        outputDataConfig: { s3OutputDataConfig: { s3Uri: `s3://${bucketName}/stop-out` } },
      })
    );

    // If the job already reached a terminal state before stop could be called, skip the stop call
    const terminal = new Set(['Completed', 'Failed', 'Stopped', 'Expired', 'PartiallyCompleted']);
    const current = await bedrock.send(new GetModelInvocationJobCommand({ jobIdentifier: createRes.jobArn }));
    if (!terminal.has(current.status as string)) {
      await bedrock.send(new StopModelInvocationJobCommand({ jobIdentifier: createRes.jobArn }));
    }

    const job = await bedrock.send(
      new GetModelInvocationJobCommand({
        jobIdentifier: createRes.jobArn,
      })
    );
    expect(['Stopping', 'Stopped', 'Failed']).toContain(job.status);
    expect(job.endTime).toBeDefined();
  });

  it('processes batch inference against S3 and produces output and manifest files', async () => {
    const inputKey = 'prompts.jsonl';
    const outputPrefix = uniqueName('batch-out');
    const inputUri = `s3://${bucketName}/${inputKey}`;
    const outputUri = `s3://${bucketName}/${outputPrefix}`;

    const records = [
      { recordId: 'rec-1', modelInput: { inputText: 'Hello Bedrock' } },
      { recordId: 'rec-2', modelInput: { inputText: 'How are you?' } },
    ];
    const jsonlContent = records.map((r) => JSON.stringify(r)).join('\n');

    await s3.send(
      new PutObjectCommand({
        Bucket: bucketName,
        Key: inputKey,
        Body: Buffer.from(jsonlContent, 'utf-8'),
      })
    );

    const jobName = uniqueName('s3-exec-job');
    const createRes = await bedrock.send(
      new CreateModelInvocationJobCommand({
        jobName,
        modelId: 'amazon.titan-text-express-v1',
        roleArn: 'arn:aws:iam::000000000000:role/BedrockBatchRole',
        inputDataConfig: { s3InputDataConfig: { s3Uri: inputUri } },
        outputDataConfig: { s3OutputDataConfig: { s3Uri: outputUri } },
      })
    );

    const jobArn = createRes.jobArn!;
    const jobId = jobArn.split('/').pop()!;

    // Poll until job reaches a terminal status
    const terminal = new Set(['Completed', 'Failed', 'Stopped', 'Expired', 'PartiallyCompleted']);
    const deadline = Date.now() + 30_000;
    let job = await bedrock.send(new GetModelInvocationJobCommand({ jobIdentifier: jobArn }));
    while (!terminal.has(job.status as string) && Date.now() < deadline) {
      await new Promise((r) => setTimeout(r, 500));
      job = await bedrock.send(new GetModelInvocationJobCommand({ jobIdentifier: jobArn }));
    }
    expect(job.status).toBe('Completed');

    // Verify output JSONL file in S3
    const outputObj = await s3.send(
      new GetObjectCommand({
        Bucket: bucketName,
        Key: `${outputPrefix}/${jobId}/${inputKey}.out`,
      })
    );
    const outputStr = await outputObj.Body!.transformToString('utf-8');
    const outputLines = outputStr.trim().split('\n');
    expect(outputLines.length).toBe(2);

    const outRecord1 = JSON.parse(outputLines[0]);
    expect(outRecord1.recordId).toBe('rec-1');
    expect(outRecord1.modelOutput).toBeDefined();

    // Verify manifest file in S3
    const manifestObj = await s3.send(
      new GetObjectCommand({
        Bucket: bucketName,
        Key: `${outputPrefix}/${jobId}/manifest.json.out`,
      })
    );
    const manifestStr = await manifestObj.Body!.transformToString('utf-8');
    const manifest = JSON.parse(manifestStr);

    expect(manifest['totalRecordCount']).toBe(2);
    expect(manifest['processedRecordCount']).toBe(2);
    expect(manifest['successRecordCount']).toBe(2);
    expect(manifest['errorRecordCount']).toBe(0);
  });
});

File: compatibility-tests/sdk-test-go/tests/bedrock_test.go
package tests

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"strings"
	"testing"
	"time"

	"floci-sdk-test-go/internal/testutil"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/bedrock"
	"github.com/aws/aws-sdk-go-v2/service/bedrock/types"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	s3types "github.com/aws/aws-sdk-go-v2/service/s3/types"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestBedrockBatchInference(t *testing.T) {
	ctx := context.Background()
	svc := testutil.BedrockClient()
	s3Client := testutil.S3Client()

	bucketName := fmt.Sprintf("go-bedrock-%d", time.Now().UnixMilli()%100000)

	_, err := s3Client.CreateBucket(ctx, &s3.CreateBucketInput{
		Bucket: aws.String(bucketName),
	})
	require.NoError(t, err, "setup: create S3 bucket")

	t.Cleanup(func() {
		list, err := s3Client.ListObjectsV2(ctx, &s3.ListObjectsV2Input{
			Bucket: aws.String(bucketName),
		})
		if err == nil {
			for _, obj := range list.Contents {
				_ = s3Client.DeleteObject(ctx, &s3.DeleteObjectInput{
					Bucket: aws.String(bucketName),
					Key:    obj.Key,
				})
			}
		}
		_ = s3Client.DeleteBucket(ctx, &s3.DeleteBucketInput{
			Bucket: aws.String(bucketName),
		})
	})

	t.Run("CreateAndGetModelInvocationJob", func(t *testing.T) {
		jobName := fmt.Sprintf("go-job-%d", time.Now().UnixMilli()%100000)
		inputUri := fmt.Sprintf("s3://%s/in.jsonl", bucketName)
		outputUri := fmt.Sprintf("s3://%s/out", bucketName)

		createOut, err := svc.CreateModelInvocationJob(ctx, &bedrock.CreateModelInvocationJobInput{
			JobName:  aws.String(jobName),
			ModelId:  aws.String("amazon.titan-text-express-v1"),
			RoleArn:  aws.String("arn:aws:iam::000000000000:role/BedrockBatchRole"),
			InputDataConfig: &types.ModelInvocationJobInputDataConfigMemberS3InputDataConfig{
				Value: types.ModelInvocationJobS3InputDataConfig{
					S3Uri: aws.String(inputUri),
				},
			},
			OutputDataConfig: &types.ModelInvocationJobOutputDataConfigMemberS3OutputDataConfig{
				Value: types.ModelInvocationJobS3OutputDataConfig{
					S3Uri: aws.String(outputUri),
				},
			},
			Tags: []types.Tag{
				{Key: aws.String("Environment"), Value: aws.String("testing")},
			},
		})
		require.NoError(t, err)
		require.NotNil(t, createOut.JobArn)
		assert.Contains(t, aws.ToString(createOut.JobArn), ":model-invocation-job/")

		getOut, err := svc.GetModelInvocationJob(ctx, &bedrock.GetModelInvocationJobInput{
			JobIdentifier: createOut.JobArn,
		})
		require.NoError(t, err)
		assert.Equal(t, aws.ToString(createOut.JobArn), aws.ToString(getOut.JobArn))
		assert.Equal(t, jobName, aws.ToString(getOut.JobName))
		assert.Equal(t, "amazon.titan-text-express-v1", aws.ToString(getOut.ModelId))
		assert.NotEmpty(t, string(getOut.Status))
	})

	t.Run("ListModelInvocationJobsWithFiltering", func(t *testing.T) {
		jobName1 := fmt.Sprintf("go-list-1-%d", time.Now().UnixMilli()%100000)
		jobName2 := fmt.Sprintf("go-list-2-%d", time.Now().UnixMilli()%100000)

		_, err := svc.CreateModelInvocationJob(ctx, &bedrock.CreateModelInvocationJobInput{
			JobName: aws.String(jobName1),
			ModelId: aws.String("amazon.titan-text-express-v1"),
			RoleArn: aws.String("arn:aws:iam::000000000000:role/BedrockBatchRole"),
			InputDataConfig: &types.ModelInvocationJobInputDataConfigMemberS3InputDataConfig{
				Value: types.ModelInvocationJobS3InputDataConfig{
					S3Uri: aws.String(fmt.Sprintf("s3://%s/1.jsonl", bucketName)),
				},
			},
			OutputDataConfig: &types.ModelInvocationJobOutputDataConfigMemberS3OutputDataConfig{
				Value: types.ModelInvocationJobS3OutputDataConfig{
					S3Uri: aws.String(fmt.Sprintf("s3://%s/out1", bucketName)),
				},
			},
		})
		require.NoError(t, err)

		_, err = svc.CreateModelInvocationJob(ctx, &bedrock.CreateModelInvocationJobInput{
			JobName: aws.String(jobName2),
			ModelId: aws.String("anthropic.claude-3-sonnet-20240229-v1:0"),
			RoleArn: aws.String("arn:aws:iam::000000000000:role/BedrockBatchRole"),
			InputDataConfig: &types.ModelInvocationJobInputDataConfigMemberS3InputDataConfig{
				Value: types.ModelInvocationJobS3InputDataConfig{
					S3Uri: aws.String(fmt.Sprintf("s3://%s/2.jsonl", bucketName)),
				},
			},
			OutputDataConfig: &types.ModelInvocationJobOutputDataConfigMemberS3OutputDataConfig{
				Value: types.ModelInvocationJobS3OutputDataConfig{
					S3Uri: aws.String(fmt.Sprintf("s3://%s/out2", bucketName)),
				},
			},
		})
		require.NoError(t, err)

		listOut, err := svc.ListModelInvocationJobs(ctx, &bedrock.ListModelInvocationJobsInput{})
		require.NoError(t, err)

		var jobNames []string
		for _, j := range listOut.InvocationJobSummaries {
			jobNames = append(jobNames, aws.ToString(j.JobName))
		}
		assert.Contains(t, jobNames, jobName1)
		assert.Contains(t, jobNames, jobName2)

		filteredOut, err := svc.ListModelInvocationJobs(ctx, &bedrock.ListModelInvocationJobsInput{
			NameContains: aws.String(jobName1),
		})
		require.NoError(t, err)

		var filteredNames []string
		for _, j := range filteredOut.InvocationJobSummaries {
			filteredNames = append(filteredNames, aws.ToString(j.JobName))
		}
		assert.Contains(t, filteredNames, jobName1)
		assert.NotContains(t, filteredNames, jobName2)
	})

	t.Run("StopModelInvocationJob", func(t *testing.T) {
		jobName := fmt.Sprintf("go-stop-%d", time.Now().UnixMilli()%100000)
		createOut, err := svc.CreateModelInvocationJob(ctx, &bedrock.CreateModelInvocationJobInput{
			JobName: aws.String(jobName),
			ModelId: aws.String("amazon.titan-text-express-v1"),
			RoleArn: aws.String("arn:aws:iam::000000000000:role/BedrockBatchRole"),
			InputDataConfig: &types.ModelInvocationJobInputDataConfigMemberS3InputDataConfig{
				Value: types.ModelInvocationJobS3InputDataConfig{
					S3Uri: aws.String(fmt.Sprintf("s3://%s/stop.jsonl", bucketName)),
				},
			},
			OutputDataConfig: &types.ModelInvocationJobOutputDataConfigMemberS3OutputDataConfig{
				Value: types.ModelInvocationJobS3OutputDataConfig{
					S3Uri: aws.String(fmt.Sprintf("s3://%s/stop-out", bucketName)),
				},
			},
		})
		require.NoError(t, err)

		_, err = svc.StopModelInvocationJob(ctx, &bedrock.StopModelInvocationJobInput{
			JobIdentifier: createOut.JobArn,
		})
		require.NoError(t, err)

		getOut, err := svc.GetModelInvocationJob(ctx, &bedrock.GetModelInvocationJobInput{
			JobIdentifier: createOut.JobArn,
		})
		require.NoError(t, err)
		assert.Contains(t, []string{"Stopping", "Stopped"}, string(getOut.Status))
		assert.NotNil(t, getOut.EndTime)
	})

	t.Run("BatchInferenceExecutionWithS3", func(t *testing.T) {
		inputKey := "prompts.jsonl"
		outputPrefix := fmt.Sprintf("go-batch-out-%d", time.Now().UnixMilli()%100000)
		inputUri := fmt.Sprintf("s3://%s/%s", bucketName, inputKey)
		outputUri := fmt.Sprintf("s3://%s/%s", bucketName, outputPrefix)

		jsonlContent := "{\"recordId\":\"rec-1\",\"modelInput\":{\"inputText\":\"Hello Bedrock\"}}\n{\"recordId\":\"rec-2\",\"modelInput\":{\"inputText\":\"How are you?\"}}"

		_, err := s3Client.PutObject(ctx, &s3.PutObjectInput{
			Bucket: aws.String(bucketName),
			Key:    aws.String(inputKey),
			Body:   bytes.NewReader([]byte(jsonlContent)),
		})
		require.NoError(t, err)

		jobName := fmt.Sprintf("go-s3-exec-%d", time.Now().UnixMilli()%100000)
		createOut, err := svc.CreateModelInvocationJob(ctx, &bedrock.CreateModelInvocationJobInput{
			JobName: aws.String(jobName),
			ModelId: aws.String("amazon.titan-text-express-v1"),
			RoleArn: aws.String("arn:aws:iam::000000000000:role/BedrockBatchRole"),
			InputDataConfig: &types.ModelInvocationJobInputDataConfigMemberS3InputDataConfig{
				Value: types.ModelInvocationJobS3InputDataConfig{
					S3Uri: aws.String(inputUri),
				},
			},
			OutputDataConfig: &types.ModelInvocationJobOutputDataConfigMemberS3OutputDataConfig{
				Value: types.ModelInvocationJobS3OutputDataConfig{
					S3Uri: aws.String(outputUri),
				},
			},
		})
		require.NoError(t, err)

		jobArn := aws.ToString(createOut.JobArn)
		jobId := jobArn[strings.LastIndex(jobArn, "/")+1:]

		getOut, err := svc.GetModelInvocationJob(ctx, &bedrock.GetModelInvocationJobInput{
			JobIdentifier: createOut.JobArn,
		})
		require.NoError(t, err)
		assert.Equal(t, string(types.ModelInvocationJobStatusCompleted), string(getOut.Status))

		// Check output JSONL in S3
		outputKey := fmt.Sprintf("%s/%s/%s.out", outputPrefix, jobId, inputKey)
		getObjOut, err := s3Client.GetObject(ctx, &s3.GetObjectInput{
			Bucket: aws.String(bucketName),
			Key:    aws.String(outputKey),
		})
		require.NoError(t, err)
		outBytes, err := io.ReadAll(getObjOut.Body)
		require.NoError(t, err)
		_ = getObjOut.Body.Close()

		lines := strings.Split(strings.TrimSpace(string(outBytes)), "\n")
		assert.Len(t, lines, 2)

		var rec1 map[string]interface{}
		err = json.Unmarshal([]byte(lines[0]), &rec1)
		require.NoError(t, err)
		assert.Equal(t, "rec-1", rec1["recordId"])
		assert.NotNil(t, rec1["modelOutput"])

		// Check manifest.json.out in S3
		manifestKey := fmt.Sprintf("%s/%s/manifest.json.out", outputPrefix, jobId)
		getManifestOut, err := s3Client.GetObject(ctx, &s3.GetObjectInput{
			Bucket: aws.String(bucketName),
			Key:    aws.String(manifestKey),
		})
		require.NoError(t, err)
		manifestBytes, err := io.ReadAll(getManifestOut.Body)
		require.NoError(t, err)
		_ = getManifestOut.Body.Close()

		var manifest map[string]interface{}
		err = json.Unmarshal(manifestBytes, &manifest)
		require.NoError(t, err)

		assert.Equal(t, float64(2), manifest["totalRecordCount"])
		assert.Equal(t, float64(2), manifest["processedRecordCount"])
		assert.Equal(t, float64(2), manifest["successRecordCount"])
		assert.Equal(t, float64(0), manifest["errorRecordCount"])
	})
}
